### PR TITLE
[http3] safer handling of CMSG (in terms of type aliasing)

### PR DESCRIPTION
**Pull request by kazuho**
When gcc 5.4.0 that comes with ubuntu16.04 is being used, we see type aliasing warnings. We do not see them when gcc 9.4.0 (ubuntu20.04) is used. Also, it is somewhat questionable if all the warnings are legit.

However, the send-side is at least hard to understand. Also, there is theoretical concern that copies involved might generate bus error due to unaligned access; there is no guarantee that `CMSG_DATA()` returns addresses that we can read and write using `=` (assignment operators) for all the types that we use.

Closes #2963.

### CODE REVIEW

File: lib/http3/common.c [hunk 1]
```diff
@@ -143,26 +148,20 @@ int h2o_quic_send_datagrams(h2o_quic_ctx_t *ctx, quicly_address_t *dest, quicly_
...
-    struct msghdr mess = {
-        .msg_name = &dest->sa,
-        .msg_namelen = quicly_get_socklen(&dest->sa),
-    };
-    if (cmsg != &cmsgbuf.hdr) {
-        mess.msg_control = &cmsgbuf.buf;
-        mess.msg_controllen = (socklen_t)((char *)cmsg - cmsgbuf.buf);
+    /* commit CMSG length */
+    if (cmsg != NULL) {
+        mess.msg_controllen = (socklen_t)((char *)cmsg - (char *)cmsgbuf.buf) + cmsg->cmsg_len;
```
Comment (i110) [inline lib/http3/common.c:158]: I'm not sure if this calculation is correct. man page says

```
Finally, the msg_controllen field of the msghdr should be set to the sum of the
CMSG_SPACE() of the length of all control messages in the buffer.
```

Are there any guarantees that this `cmsg - start + cmsg->cmsg_len` value is equal to that sum of CMSG_SPACE? Even if so, I feel it's more straightforward
to do `controlllen += CMSG_SPACE(sizeof(value));` in PUSH_CMSG and set it to mess.msg_controlllen here.

**Review by i110 — APPROVED** [review]
Thank you for the change! The code is much easier to read now.

File: lib/http3/common.c (hunk 1 above)
Comment (kazuho) [reply lib/http3/common.c:158]: That's a really keen observation. I've found [a good visualization in RFC 3542](https://datatracker.ietf.org/doc/html/rfc3542#page-62) explaining what these values mean, and it does seem that it explicitly permits such structure (quote: _While sending an application may or may not include padding at the end of last ancillary data in msg_controllen and implementations must accept both as valid._).

Though, I agree with you that we have an inconsistency here and that we can do better. Will see what I can do.

Comment (i110) [reply lib/http3/common.c]: Yeah, the sentence you quoted explains all 👍

**Pull request merged**